File: python/src/agent_framework/crew/crew_orchestrator.py
"""Crew orchestrator: runs Crew with optional message bus."""

from typing import Any
from agent_framework.crew.models import CrewContext
from agent_framework.crew.message_bus import MessageBus


class CrewRunOptions:
    """Options for a crew run."""

    def __init__(
        self,
        thread_id: str = "default",
        initial_input: dict[str, Any] | None = None,
    ):
        self.thread_id = thread_id
        self.initial_input = initial_input or {}


class CrewRunResult:
    """Result of a crew run."""

    def __init__(
        self,
        outputs: dict[str, Any],
        completed_task_ids: list[str],
    ):
        self.outputs = outputs
        self.completed_task_ids = completed_task_ids


class CrewOrchestrator:
    """Orchestrates a Crew (agents + tasks + tools)."""

    def __init__(self, context: CrewContext, message_bus: MessageBus | None = None) -> None:
        self._context = context
        self._message_bus = message_bus

    async def run(self, options: CrewRunOptions | None = None) -> CrewRunResult:
        opts = options or CrewRunOptions()
        outputs: dict[str, Any] = dict(opts.initial_input)
        completed_task_ids: list[str] = []
        crew = self._context.crew
        agents = self._context.agents
        tasks = self._context.tasks
        order = crew.task_ids

        for task_id in order:
            task = tasks.get(task_id)
            if not task:
                continue
            agent_id = task.agent_id or (crew.agent_ids[0] if crew.agent_ids else "")
            agent = agents.get(agent_id) if agent_id else None
            if not agent:
                continue

            if self._message_bus:
                await self._message_bus.publish(
                    f"crew:{crew.id}",
                    agent_id,
                    {"type": "task_start", "task_id": task_id, "task_description": task.description},
                )

            result = await self._execute_task(task, agent.id, outputs)
            outputs[task_id] = result
            completed_task_ids.append(task_id)

            if self._message_bus:
                await self._message_bus.publish(
                    f"crew:{crew.id}",
                    agent_id,
                    {"type": "task_done", "task_id": task_id, "result": result},
                )

        return CrewRunResult(outputs=outputs, completed_task_ids=completed_task_ids)

    async def _execute_task(
        self,
        task: Any,
        agent_id: str,
        context: dict[str, Any],
    ) -> Any:
        if getattr(task, "output", None) is not None:
            return task.output
        return {"input": getattr(task, "input", None), "context": context, "agent_id": agent_id}

File: packages/crew/src/crew-orchestrator.ts
import type { MessageBus } from "./message-bus.js";
import type { AgentDefinition, CrewDefinition, CrewContext, Task } from "@agent-framework/core";
import type { HumanInTheLoopOptions, InterceptHandler } from "@agent-framework/runtime";

export interface CrewRunOptions {
  threadId?: string;
  initialInput?: Record<string, unknown>;
  /** Override HITL for this run */
  hitl?: HumanInTheLoopOptions | null;
}

export interface CrewRunResult {
  outputs: Record<string, unknown>;
  completedTaskIds: string[];
  /** True if terminated by HITL */
  terminatedByHuman?: boolean;
}

/**
 * Crew orchestrator: runs a Crew (agents + tasks + tools) with optional message bus, delegation, and HITL.
 */
export class CrewOrchestrator {
  constructor(
    private context: CrewContext,
    private messageBus?: MessageBus,
    private hitlOptions?: HumanInTheLoopOptions
  ) {}

  private async runCrewIntercept(
    context: { threadId: string; taskId: string; taskDescription: string; outputs: Record<string, unknown>; step: number },
    hitl: HumanInTheLoopOptions
  ): Promise<{ decision: "approve" | "skip" | "terminate" | "modify"; modifiedOutputs?: Record<string, unknown> }> {
    const { handler, webhookUrl, timeoutMs = 30_000 } = hitl;
    const ctx = {
      threadId: context.threadId,
      nodeId: context.taskId,
      step: context.step,
      state: context.outputs,
      proposedAction: `task:${context.taskId}`,
      proposedOutput: undefined,
      timestamp: Date.now() / 1000,
    };
    if (webhookUrl) {
      try {
        const controller = new AbortController();
        const id = setTimeout(() => controller.abort(), timeoutMs);
        const res = await fetch(webhookUrl, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(ctx),
          signal: controller.signal,
        });
        clearTimeout(id);
        const body = (await res.json()) as { decision: string; modifiedState?: Record<string, unknown> };
        return { decision: body.decision as "approve" | "skip" | "terminate" | "modify", modifiedOutputs: body.modifiedState };
      } catch {
        return { decision: "skip" };
      }
    }
    if (handler) {
      try {
        const resp = await Promise.race([
          Promise.resolve(handler(ctx)),
          new Promise<never>((_, reject) => setTimeout(() => reject(new Error("HITL timeout")), timeoutMs)),
        ]);
        return { decision: resp.decision, modifiedOutputs: resp.modifiedState };
      } catch {
        return { decision: "skip" };
      }
    }
    return { decision: "approve" };
  }

  /**
   * Run crew in sequential order; tasks are executed by assigned agents.
   */
  async run(options: CrewRunOptions = {}): Promise<CrewRunResult> {
    const { threadId = "default", initialInput = {}, hitl: hitlOverride } = options;
    const hitl = hitlOverride ?? this.hitlOptions;
    const interceptAlways = hitl?.mode === "ALWAYS";
    const outputs: Record<string, unknown> = { ...initialInput };
    const completedTaskIds: string[] = [];
    const { crew, agents, tasks } = this.context;
    const order = crew.executionOrder === "sequential" ? crew.taskIds : [...crew.taskIds];

    for (let step = 0; step < order.length; step++) {
      const taskId = order[step];
      const task = tasks.get(taskId);
      if (!task) continue;
      const agentId = task.agentId ?? crew.agentIds[0];
      const agent = agents.get(agentId);
      if (!agent) continue;

      if (interceptAlways && hitl) {
        const { decision, modifiedOutputs } = await this.runCrewIntercept(
          { threadId, taskId, taskDescription: task.description, outputs, step },
          hitl
        );
        if (decision === "terminate") {
          return { outputs, completedTaskIds, terminatedByHuman: true };
        }
        if (decision === "modify" && modifiedOutputs) {
          Object.assign(outputs, modifiedOutputs);
        }
      }

      if (this.messageBus) {
        await this.messageBus.publish({
          channel: `crew:${crew.id}`,
          senderId: agentId,
          payload: { type: "task_start", taskId, taskDescription: task.description },
        });
      }

      const result = await this.executeTask(task, agent, outputs);
      outputs[taskId] = result;
      completedTaskIds.push(taskId);

      if (this.messageBus) {
        await this.messageBus.publish({
          channel: `crew:${crew.id}`,
          senderId: agentId,
          payload: { type: "task_done", taskId, result },
        });
      }
    }

    return { outputs, completedTaskIds };
  }

  private async executeTask(
    task: Task,
    agent: AgentDefinition,
    context: Record<string, unknown>
  ): Promise<unknown> {
    if (task.output !== undefined) return task.output;
    return { input: task.input, context, agentId: agent.id };
  }
}
